src/js/experiments/snippetFeedback.js

```diff
@@ -0,0 +1,250 @@
+      this.wasThisHelpfulContainer.appendChild(container);
+
+      // Here we setup the document event listeners that handle hiding the form
+      // when the user clicks outside of it.
+      //
+      // Why do we use a setTimeout with a timer of 0?
+      // -> Because there is a race condition with the adding of the click
+      // listeners to the document that immediately close the form when it opens
+      //
+      // Setting it to 0 causes the inner callback to be run asynchronously, adding it to the eventqueue
+      // so that it executes after the form being opened.
+      setTimeout(() => {
+        const outsideClickListener = (event) => {
+          if (!this.feedbackForm.contains(event.target)) {
+            this.wasThisHelpfulContainer.removeChild(this.feedbackForm);
+            this.feedbackForm = null;
+            removeClickListener();
+          }
+        };
+
+        const removeClickListener = () => {
+          document.removeEventListener('click', outsideClickListener);
+        };
+
+        document.addEventListener('click', outsideClickListener);
+      }, 0);
+    }
+  }
+
+  /**
+   * Render the char count and set it's result on the respective element.
+   * */
+  renderCharCount(charCount) {
+    let charCountLimited =
+      charCount > this.maxCharCount ? this.maxCharCount : charCount;
+    const el = `
+        <div class="charCountRow">
+          <span>${charCountLimited} / ${this.maxCharCount}</span>
+        </div>`;
+    this.currCharCountElement.innerHTML = el;
+  }
+
+  // -- Trackers --
+  // These are responsible for sending data based on user input
+  // to amplitude.
+
+  /**
+   * trackYesOrNoButton sends a trackAction to aplitude.
+   * actions are: "docs-snippet-helpful-no"  and "docs-snippet-helpful-yes"
+   * */
+  trackYesOrNoButton(yesOrNoString) {
+    window.AnalyticsClient.trackAction(
+      `docs-snippet-helpful-${yesOrNoString}`,
+      {
+        originatingSnippet: this.snippetElement.textContent,
+        location: window.location.pathname,
+        timeOfButtonClick: Date.now(),
+      },
+    );
+  }
+
+  /**
+   * `trackFormSubmission` invokes trackAction with a payload of the original
+   * snippet, the feedback of the user, and the timeOfFormSubmission.
+   * */
+  trackFormSubmission(formContent) {
+    window.AnalyticsClient.trackAction(`docs-snippet-helpful-form-submission`, {
+      originatingSnippet: this.snippetElement.textContent,
+      feedback: formContent,
+      location: window.location.pathname,
+      wasThisHelpful: this.wasThisHelpful,
+      // for diffing this against the time that the trackYesOrNoButton was clicked
+      // as well as possible the time the copy code button was clicked (TODO add that)
+      timeOfFormSubmission: Date.now(),
+    });
+  }
+
+  // -- Helpers --
+
+  /**
+   * Construct a dom element, setting classes, text content, class, onclick etc.
+   * */
+  makeElement({ kind, text, className, onClick }) {
```

Comment:
maybe some of these class methods like this should be made private?

Reply:
We do not have access to public/private keywords in the version of JavaScript we are using.

Reply:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Private_class_fields#browser_compatibility am I missing something?

if we're not able to, can we follow convention of using _ before the method name? https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Working_With_Private_Class_Features